root.go

```diff
@@ -27,28 +20,76 @@ func (r MerkleRoot) MustSum() uint64 {
 // Sum returns the uint64 sum of the merkle root, it checks the length of the
 // merkle root and if it is no the same as the size of the SMST's expected
 // root hash it will return an error.
-func (r MerkleRoot) Sum() (uint64, error) {
-	if len(r)%SmtRootSizeBytes == 0 {
-		return 0, fmt.Errorf("root#sum: not a merkle sum trie")
+func (root MerkleSumRoot) Sum() (uint64, error) {
+	if err := root.validateBasic(); err != nil {
+		return 0, err
 	}
 
-	firstSumByteIdx, firstCountByteIdx := getFirstMetaByteIdx([]byte(r))
+	return root.sum(), nil
+}
+
+// MustCount returns the uint64 count of the merkle root, a cryptographically secure
+// count of the number of non-empty leafs in the tree. It panics if the root length
+// is invalid.
+func (root MerkleSumRoot) MustCount() uint64 {
+	count, err := root.Count()
+	if err != nil {
+		panic(err)
+	}
 
-	var sumBz [sumSizeBytes]byte
-	copy(sumBz[:], []byte(r)[firstSumByteIdx:firstCountByteIdx])
-	return binary.BigEndian.Uint64(sumBz[:]), nil
+	return count
 }
 
 // Count returns the uint64 count of the merkle root, a cryptographically secure
-// count of the number of non-empty leafs in the tree.
-func (r MerkleRoot) Count() uint64 {
-	if len(r)%SmtRootSizeBytes == 0 {
-		panic("root#sum: not a merkle sum trie")
+// count of the number of non-empty leafs in the tree. It returns an error if the
+// root length is invalid.
+func (root MerkleSumRoot) Count() (uint64, error) {
+	if err := root.validateBasic(); err != nil {
+		return 0, err
 	}
 
-	_, firstCountByteIdx := getFirstMetaByteIdx([]byte(r))
+	return root.count(), nil
+}
+
+// DigestSize returns the length of the digest portion of the root.
+func (root MerkleSumRoot) DigestSize() int {
+	return len(root) - countSizeBytes - sumSizeBytes
+}
+
+// HasDigestSize returns true if the root digest size is the same as
+// that of the size of the given hasher.
+func (root MerkleSumRoot) HasDigestSize(size int) bool {
+	return root.DigestSize() == size
+}
 
-	var countBz [countSizeBytes]byte
-	copy(countBz[:], []byte(r)[firstCountByteIdx:])
-	return binary.BigEndian.Uint64(countBz[:])
+// validateBasic returns an error if the root digest size is not a power of two.
```

Comment:
NIT: Move this comment INSIDE the function and keep the godoc more general OR call out a power of two as an explicit example.

Reply:
I'm either not quite following or am just unable to think of an alternative godoc comment that satisfies your suggestion. Do you have a concrete example in mind (i.e., `validateBasic returns an error if ...`)?